actors/builtin/verifreg/verified_registry_actor.go

```diff
@@ -318,3 +319,120 @@ func (a Actor) RestoreBytes(rt runtime.Runtime, params *RestoreBytesParams) *abi
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to resolve client address %v to ID address", params.VerifiedClientToRemove)
+	verifier1, err := builtin.ResolveToIDAddr(rt, params.VerifierRequest1.Verifier)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to resolve verifier address %v to ID address", params.VerifierRequest1.Verifier)
+	verifier2, err := builtin.ResolveToIDAddr(rt, params.VerifierRequest2.Verifier)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to resolve verifier address %v to ID address", params.VerifierRequest2.Verifier)
+
+	if verifier1 == verifier2 {
+		rt.Abortf(exitcode.ErrIllegalArgument, "need two different verifiers to send remove datacap request got %v and %v that are the same accounts", params.VerifierRequest1.Verifier,
+			params.VerifierRequest2.Verifier)
+	}
+
+	var st State
+	rt.StateReadonly(&st)
+
+	rt.ValidateImmediateCallerIs(st.RootKey)
+
+	var removedDataCapAmount = params.DataCapAmountToRemove // amount of datacap removed
+	rt.StateTransaction(&st, func() {
+		// validate the client is a verified client and get the current DataCap the client holds
+		verifiedClients, err := adt.AsMap(adt.AsStore(rt), st.Verifiers, builtin.DefaultHamtBitwidth)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load verified clients")
+		var preDataCap DataCap // amount of datacap the client currently holds
+		isVerifiedClient, err := verifiedClients.Get(abi.AddrKey(client), &preDataCap)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to get verified client %v", params.VerifiedClientToRemove)
+		if !isVerifiedClient {
+			rt.Abortf(exitcode.ErrNotFound, "%v is not a verified client", params.VerifiedClientToRemove)
+		}
+
+		// validate requesters are verifiers
+		if isVerifier, code, err := isVerifier(rt, st, verifier1); !isVerifier {
+			rt.Abortf(code, err.Error())
+		}
+		if isVerifier, code, err := isVerifier(rt, st, verifier2); !isVerifier {
+			rt.Abortf(code, err.Error())
+		}
+
+		// get the remove datacap proposal IDs for  the verifiers
+		proposalIDs, err := adt.AsMap(adt.AsStore(rt), st.RemoveDataCapProposalIDs, builtin.DefaultHamtBitwidth)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load datacap removal proposal ids")
+
+		var verifier1ID RmDcProposalID
+		idExists, err := proposalIDs.Get(abi.NewAddrPairKey(verifier1, client), &verifier1ID)
+		if !idExists { //initialize with 0
+			verifier1ID = RmDcProposalID{proposalID: 0}
+			err = proposalIDs.Put(abi.NewAddrPairKey(verifier1, client), &verifier1ID)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add remove datacap proposal id for the verifier")
+		}
+
+
+		var verifier2ID RmDcProposalID
+		idExists, err = proposalIDs.Get(abi.NewAddrPairKey(verifier2, client), &verifier2ID)
+		if !idExists { //initialize with 0
+			verifier2ID = RmDcProposalID{proposalID: 0}
```

Comment:
this should be a subfunction

Reply:
we can do the increment at the same time too

Reply:
wouldn't it be better if we wait for the rest of the checks and computations done before we do the two puts? to save some gas?

Reply:
How would it save gas?  If the message succeeds this needs to happen so no savings assuming success.  Are you thinking of saving gas in error cases?  The sender pays the whole gas limit if it hits an abort so no savings there.

Reply:
Ah ha- that makes sense. I somehow was thinking that we only charge gas  used upon to the  error point.